File: app/src/main/java/org/wikipedia/dataclient/page/PageSummary.java
package org.wikipedia.dataclient.page;

import android.location.Location;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.annotations.JsonAdapter;
import com.google.gson.annotations.SerializedName;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.dataclient.WikiSite;
import org.wikipedia.page.GeoTypeAdapter;
import org.wikipedia.page.Namespace;
import org.wikipedia.page.Page;
import org.wikipedia.page.PageProperties;
import org.wikipedia.page.PageTitle;
import org.wikipedia.util.UriUtil;

/**
 * Represents a summary of a page, useful for page previews.
 */
@SuppressWarnings("unused")
public class PageSummary {
    public static final String TYPE_STANDARD = "standard";
    public static final String TYPE_DISAMBIGUATION = "disambiguation";
    public static final String TYPE_MAIN_PAGE = "mainpage";
    public static final String TYPE_NO_EXTRACT = "no-extract";

    @Nullable private String type;
    @Nullable private Titles titles;
    @Nullable private NamespaceContainer namespace;
    @Nullable private String extract;
    @Nullable @SerializedName("extract_html") private String extractHtml;
    @Nullable private String description;
    @Nullable @SerializedName("description_source") private String descriptionSource;
    @Nullable private Thumbnail thumbnail;
    @Nullable @SerializedName("originalimage") private Thumbnail originalImage;
    @Nullable private String lang;
    private int pageid;

    private long revision;
    @Nullable @JsonAdapter(GeoTypeAdapter.class) private Location coordinates;
    @Nullable private String timestamp;
    @SerializedName("wikibase_item") @Nullable private String wikiBaseItem;

    public PageSummary() { }

    public PageSummary(@NonNull String displayTitle, @NonNull String prefixTitle, @Nullable String description,
                       @Nullable String extract, @Nullable String thumbnail, @NonNull String lang) {
        this.titles = new Titles(prefixTitle, displayTitle);
        this.description = description;
        this.extract = extract;
        this.thumbnail = new Thumbnail(thumbnail, 0, 0);
        this.lang = lang;
    }

    public Page toPage(PageTitle title) {
        return new Page(adjustPageTitle(title), new PageProperties(this));
    }

    private PageTitle adjustPageTitle(PageTitle title) {
        PageTitle newTitle = title;
        if (titles != null && titles.canonical != null) {
            newTitle = new PageTitle(titles.canonical, title.getWikiSite(), title.getThumbUrl());
            newTitle.setFragment(title.getFragment());
        }
        newTitle.setDescription(description);
        return newTitle;
    }

    @NonNull
    public String getApiTitle() {
        return StringUtils.defaultString(titles != null ? titles.canonical : null);
    }

    // TODO: Make this return CharSequence, and automatically convert from HTML.
    @NonNull
    public String getDisplayTitle() {
        return StringUtils.defaultString(titles != null ? titles.display : null);
    }

    @NonNull
    public Namespace getNamespace() {
        return namespace == null ? Namespace.MAIN : Namespace.of(namespace.id());
    }

    @NonNull
    public String getType() {
        return TextUtils.isEmpty(type) ? TYPE_STANDARD : type;
    }

    @Nullable
    public String getExtract() {
        return extract;
    }

    @Nullable
    public String getExtractHtml() {
        return extractHtml;
    }

    @Nullable
    public String getThumbnailUrl() {
        return thumbnail == null ? null : thumbnail.getUrl();
    }

    public int getThumbnailWidth() {
        return thumbnail == null ? 0 : thumbnail.getWidth();
    }

    public int getThumbnailHeight() {
        return thumbnail == null ? 0 : thumbnail.getHeight();
    }

    public void setDescription(@Nullable String description) {
        this.description = description;
    }

    @Nullable
    public String getDescription() {
        return description;
    }

    @NonNull
    public String getDescriptionSource() {
        return StringUtils.defaultString(descriptionSource);
    }

    @Nullable
    public String getOriginalImageUrl() {
        return originalImage == null ? null : originalImage.getUrl();
    }

    @NonNull
    public PageTitle getPageTitle(@NonNull WikiSite wiki) {
        return new PageTitle(getApiTitle(), wiki, getThumbnailUrl(), getDescription(), getDisplayTitle(), getExtract());
    }

    public int getPageId() {
        return pageid;
    }

    @NonNull
    public String getLang() {
        return StringUtils.defaultString(lang);
    }

    private static class Thumbnail {
        private final String source;
        private final int width;
        private final int height;

        Thumbnail(@Nullable String source, int width, int height) {
            this.source = source;
            this.width = width;
            this.height = height;
        }

        public String getUrl() {
            return source;
        }

        public int getWidth() {
            return width;
        }

        public int getHeight() {
            return height;
        }
    }

    private static class NamespaceContainer {
        private int id;
        @Nullable private String text;

        public int id() {
            return id;
        }
    }

    private static class Titles {
        @Nullable private String canonical;
        @Nullable private String display;

        Titles(@Nullable String canonical, @Nullable String display) {
            this.canonical = canonical;
            this.display = display;
        }
    }

    @Override @NonNull public String toString() {
        return getDisplayTitle();
    }

    public long getRevision() {
        return revision;
    }

    @Nullable
    public Location getGeo() {
        return coordinates;
    }

    @Nullable
    public String getTimestamp() {
        return StringUtils.defaultString(timestamp);
    }

    @Nullable
    public String getWikiBaseItem() {
        return StringUtils.defaultString(wikiBaseItem);
    }

    @Nullable
    public String getLeadImageName() {
        if (getThumbnailUrl() == null) {
            return null;
        }
        return UriUtil.getFilenameFromUploadUrl(getThumbnailUrl());
    }
}

File: app/src/main/java/org/wikipedia/dataclient/page/PageSummary.kt
package org.wikipedia.dataclient.page

import android.location.Location
import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.wikipedia.dataclient.WikiSite
import org.wikipedia.json.LocationSerializer
import org.wikipedia.page.Namespace
import org.wikipedia.page.Page
import org.wikipedia.page.PageProperties
import org.wikipedia.page.PageTitle
import org.wikipedia.util.UriUtil.getFilenameFromUploadUrl

@Parcelize
@Serializable
open class PageSummary(
    val namespace: NamespaceContainer? = null,
    var titles: Titles? = null,
    var lang: String = "",
    var thumbnail: Thumbnail? = null,
    var extract: String? = null,
    var description: String? = null,
    @SerialName("originalimage") private val originalImage: Thumbnail? = null,
    @SerialName("wikibase_item") val wikiBaseItem: String? = null,
    @SerialName("extract_html") val extractHtml: String? = null,
    @SerialName("description_source") val descriptionSource: String = "",
    @Serializable(with = LocationSerializer::class) val geo: Location? = null,
    val type: String = TYPE_STANDARD,
    @SerialName("pageid") val pageId: Int = 0,
    val revision: Long = 0L,
    val timestamp: String = "",
    val views: Long = 0,
    private val rank: Long = 0,
    @SerialName("view_history") val viewHistory: List<ViewHistory>? = null
) : Parcelable {

    val thumbnailUrl get() = thumbnail?.source
    val thumbnailWidth get() = thumbnail?.width ?: 0
    val thumbnailHeight get() = thumbnail?.height ?: 0
    val originalImageUrl get() = originalImage?.source
    val apiTitle get() = titles?.canonical.orEmpty()

    // TODO: Make this return CharSequence, and automatically convert from HTML.
    val displayTitle get() = titles?.display.orEmpty()

    val leadImageName get() = thumbnailUrl?.let { getFilenameFromUploadUrl(it) }

    val ns: Namespace get() = if (namespace == null) Namespace.MAIN else Namespace.of(namespace.id)

    constructor(displayTitle: String, prefixTitle: String, description: String?,
                extract: String?, thumbnail: String?, lang: String) : this(
        titles = Titles(prefixTitle, displayTitle), description = description, extract = extract,
        thumbnail = Thumbnail(thumbnail, 0, 0), lang = lang
    )

    fun toPage(title: PageTitle): Page {
        return Page(adjustPageTitle(title), pageProperties = PageProperties(this))
    }

    private fun adjustPageTitle(title: PageTitle): PageTitle {
        var newTitle = title
        if (titles?.canonical != null) {
            newTitle = PageTitle(titles?.canonical, title.wikiSite, title.thumbUrl)
            newTitle.fragment = title.fragment
        }
        newTitle.description = description
        return newTitle
    }

    fun getPageTitle(wiki: WikiSite): PageTitle {
        return PageTitle(apiTitle, wiki, thumbnailUrl, description, displayTitle, extract)
    }

    override fun toString(): String {
        return displayTitle
    }

    @Parcelize
    @Serializable
    data class NamespaceContainer(val id: Int = 0, val text: String = "") : Parcelable

    @Parcelize
    @Serializable
    class Titles(val canonical: String?, val display: String?) : Parcelable

    @Parcelize
    @Serializable
    class Thumbnail(val source: String?, val width: Int, val height: Int) : Parcelable

    @Parcelize
    @Serializable
    class ViewHistory(val date: String, val views: Float) : Parcelable

    companion object {
        const val TYPE_STANDARD = "standard"
        const val TYPE_DISAMBIGUATION = "disambiguation"
        const val TYPE_MAIN_PAGE = "mainpage"
        const val TYPE_NO_EXTRACT = "no-extract"
    }
}
